src/helpers.rs

```diff
@@ -912,30 +912,57 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
         })
     }
 
-    /// Read a sequence of bytes until the first null terminator.
-    fn read_c_str<'a>(&'a self, ptr: Pointer<Option<Provenance>>) -> InterpResult<'tcx, &'a [u8]>
-    where
-        'tcx: 'a,
-        'mir: 'a,
-    {
+    /// Generic version of strlen to support multiple character sizes.
```

Comment:
I think you can have a generic read_str, no need to add strlen functions here.

The function should take
- a `TyAndLayout` describing the type to do the loads with
- a generic `T: TryFrom<u128>` to use for the vector that it returns

Reply:
The point is that we only need `strlen` for implementing `strlen` and `strcpy`.

Furthermore `wcslen` only needs `wchar_t_strlen` and not some `read_w_char_t_str`.

Reply:
Sure, but we do need `read_wide_str`, and if we can just make that generic then it can be used by `wcslen` and the existing os_str helpers, so we have less code to maintain. :)